GSIM and GLIM with test cases (FINERACT-603)

rrpawar96 opened the pull request:
## Description
fixed some bugs in gsim_glim and created the Integration test for it.

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Commit message starts with the issue number from 
        https://github.com/apache/fineract/pull/527

- [ ] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [ ] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [ ] Integration tests have been created/updated for verifying the changes made.

- [ ] All Integrations tests are passing with the new commits.

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

rrpawar96 committed:
GLIM_GSIM enhancement (FINERACT-603)

maektwain commented:
@rrpawar96 So this is more like a checkStyle issue , you need to install the check style plugin in your IDE and use the checkstyle config , this seems like a most common, issue though in eclipse I verified it is unable to apply the check style fix automatically

@vorburger most probably we need some service which can do on its own, internally packaged or built inside 

For now , @rrpawar96  you can use this, https://checkstyle.sourceforge.io/cmdline.html#Command_line_usage

https://stackoverflow.com/questions/8409074/how-can-i-easily-fix-checkstyle-errors

Sadly in my case ,I had to do this manually.

rahulg963 commented:
@rrpawar96 any specific reason for using System.out.print() instead of a logger?

xurror commented:
> @rrpawar96 So this is more like a checkStyle issue , you need to install the check style plugin in your IDE and use the checkstyle config , this seems like a most common, issue though in eclipse I verified it is unable to apply the check style fix automatically
> 
> @vorburger most probably we need some service which can do on its own, internally packaged or built inside
> 
> For now , @rrpawar96 you can use this, https://checkstyle.sourceforge.io/cmdline.html#Command_line_usage
> 
> https://stackoverflow.com/questions/8409074/how-can-i-easily-fix-checkstyle-errors
> 
> Sadly in my case ,I had to do this manually.

Can you file a ticket for this?

vorburger commented:
Hello every one, could you elaborate on what exactly you are looking for here? Do you want to...

1. Just run Checkstyle on the CLI?  You don't have to use https://checkstyle.sourceforge.io/cmdline.html#Command_line_usage for that, you can just use Gradle, no? If I'm missing something, please clarify.

2. Have Checkstyle in Eclipse? You can use https://checkstyle.org/eclipse-cs/ for that! You should be able just load [our checkstyle.xml](https://github.com/apache/fineract/blob/develop/fineract-provider/config/checkstyle/) into it. I have not tried that yet, but I don't see why that would not work. I'd love you to try it, and then contribute a PR documenting how to do this to https://github.com/apache/fineract#checkstyle. 

BTW importing config/fineractdev-eclipse-preferences.epf as documented on our README normally automatically does much of what our Checkstyle configuration requires for you already. If you just do that, you may not even need to install and set up the Eclipse Checkstyle plugin? Try both, and see what works best for you, and share what you learnt here and on the dev list.

vorburger commented:
> contribute a PR documenting how to do this

==> #747

rrpawar96 commented:
> @rrpawar96 any specific reason for using System.out.print() instead of a logger?

Not really, but it is used because till now all test cases scripts has used  System.out.print() instead of a logger and also it helps for quick debugging, so used it.

thesmallstar commented:
@rrpawar96 hello, I suggest you use logger, use of System.out would lead to checkstyle error after #733 is merged.
-thanks

vorburger commented:
@rrpawar96 this still seems to have build failures, did you notice? The build logs on Travis CI are not accessible right now (it says "There was an error while trying to fetch the log."), probably due to https://www.traviscistatus.com showing that Travis CI is currently undergoing some maintenance... will you check out later (tomorrow) what the failures are, and fix? Then we can re-review and hopefully soon merge this. Also, once this is all good, the commits should probably be "squashed" into a single commit.

rrpawar96 commented:
Yes @vorburger, I have fixed all the CheckStyle and SpotBugs issues, now the Travis build is failing in Integration test. (3 failed test cases are: ClientLoanIntegrationTest.testLoanCharges_DISBURSEMENT_FEE
ClientLoanIntegrationTest.testLoanCharges_INSTALMENT_FEE
ClientLoanIntegrationTest.testLoanCharges_SPECIFIED_DUE_DATE_FEE) and failed with the cause: totalLoan paramter is unsupported.

vorburger commented:
@rrpawar96 what you need to do is figure out if these integration test failures are due to the changes in this PR, or arbitrarily failing tests... does running `./gradlew clean integrationTest` work for you locally? If no, figure out how to fix it locally. If yes, and based on https://issues.apache.org/jira/browse/FINERACT-708?jql=text%20~%20%22ClientLoanIntegrationTest%22 it would look like these may just "flaky". In that case, do a rebase, and see if a new build passes. If not, wait a day, close and re-open the issue, and see if it's better.

rrpawar96 commented:
@vorburger , Thankyou for ur feedback and as I followed it and found out that the cause for failing  test cases of previous commit was this PR, so I fixed it and rebased the code and locally tested it. So locally the all integration test build got passed so after pushing it, now on Travis unfortunately, FixedDepositTest got failed.

rrpawar96 commented:
@vorburger , should I squash all the commits and try again by running on Travis ?

vorburger commented:
Yeah, give it a try...

>

vorburger commented:
FTR the current build failure is definitely due to a flaky IT, see [FINERACT-887](https://jira.apache.org/jira/browse/FINERACT-887).. just (squash and) keep closing and re-opening this PR.

PS: @rahulg963 any future contributions to fix these "flaky" ITs would be very, very, welcome; see [FINERACT-850](https://jira.apache.org/jira/browse/FINERACT-850)!

rrpawar96 commented:
> FTR the current build failure is definitely due to a flaky IT, see [FINERACT-887](https://jira.apache.org/jira/browse/FINERACT-887).. just (squash and) keep closing and re-opening this PR.

Yes, now code has been again rebased and squashed all the commits into one, now will try to pass the PR by closing and re-opening it.

rrpawar96 commented:
@vorburger, and finally build got passed by Travis. 😊

vorburger commented:
> @vorburger, and finally build got passed by Travis. blush

That was most likely more of a coincidence actually... FYI there were 2 flaky tests, and sometimes it passes, and sometimes it failed.. :smile: Following #764 and #761 this would always pass the build (or not at least not fail because of integration tests) if you rebased this now. 

I'm assuming that this PR fixes https://issues.apache.org/jira/browse/FINERACT-603, is this correct? I'd like to assign that issue to you instead of myself (or formerly Nazeer Shaik ) in JIRA to you, so that you "get credit" whenever someone will create Release Notes, but I cannot find you in JIRA - what is your ID on https://issues.apache.org/jira? If you do not have an account yet, perhaps this would be a great time to create one?

I'll try to make time for a more details rewiew of this (big!) PR hopefully in the coming days.

rrpawar96 commented:
> I'm assuming that this PR fixes https://issues.apache.org/jira/browse/FINERACT-603, is this correct? I'd like to assign that issue to you instead of myself (or formerly Nazeer Shaik ) in JIRA to you, so that you "get credit" whenever someone will create Release Notes, but I cannot find you in JIRA - what is your ID on https://issues.apache.org/jira? If you do not have an account yet, perhaps this would be a great time to create one?

Yes, @vorburger, that is the one https://issues.apache.org/jira/browse/FINERACT-603 which is fixed by this PR.
My JIRA Id is : rrpawar

Thanks.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -86,470 +93,715 @@
+
+                              if (gsimRepository.count() != 0) {
+                                   // Child-Not an empty table check
+                                   gsimAccount = gsimRepository.findOneByIsAcceptingChildAndApplicationId(true,applicationId);
```
@rrpawar96 if client/front end application generates same applicationID for two applications- it would mix up the child applications.  Although the chances are very low, we should see if we can synchronize access to submitApplication() through its parent submitGSIMApplication().

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -86,470 +93,715 @@
+
+          JsonArray gsimApplications = command.arrayOfParameterNamed("clientArray");
+
+          for (JsonElement gsimApplication : gsimApplications) {
```
@rrpawar96  you may want to create an object and lock this loop until it completes processing. If you are able to do it, you can remove applicationID param altogether.

vorburger commented on fineract-provider/src/main/resources/META-INF/persistence.xml:
```diff
@@ -28,6 +28,8 @@
 	<persistence-unit name="jpa-pu" transaction-type="RESOURCE_LOCAL">
 		<provider>org.apache.openjpa.persistence.PersistenceProviderImpl</provider>
 	<properties>
+		<property name="openjpa.jdbc.EagerFetchMode" value="none"/>
+		<property name="openjpa.jdbc.SubclassFetchMode" value="none"/>
```
These two OpenJPA configuration changes here (`EagerFetchMode` & `SubclassFetchMode`) strike me as major data access changes that perhaps should not be "hidden" within this PR.. would you be willing to raise a new JIRA and a separate PR and remove this from here so that it can be reviewed and discussed in its own right?

nikpawar89 replied:
i agree with @vorburger

rrpawar96 replied:
Comment has been addressed and new PR has been created: https://github.com/apache/fineract/pull/777

vorburger commented:
> I'll try to make time for a more details rewiew of this (big!) PR hopefully in the coming days.

I glanced over this today, and have 1 major feedback re. the OpenJPA configuration changes proposed here (re. `EagerFetchMode` & `SubclassFetchMode`) - personally I would much prefer to see this brought up completely separately, and not "tucked away" hidden inside this PR....

Other than that, this is a (big, thank you!) functional change which others than me would be better place to review; I've just replied on the "Re: Re-opening Pull Request and Writing Test Cases for https://github.com/apache/fineract/pull/527" thread on the dev list with a request for this.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/accountdetails/service/AccountDetailsReadPlatformServiceJpaRepositoryImpl.java:
```diff
@@ -131,6 +162,16 @@ public AccountSummaryCollectionData retrieveGroupAccountDetails(final Long group
+    @Override
+        public List<LoanAccountSummaryData> retrieveLoanAccountDetailsByGroupIdAndGlimAccountNumber(final Long groupId,final String glimAccount) {
+        final LoanAccountSummaryDataMapper rm = new LoanAccountSummaryDataMapper();
+        final String loanWhereClauseForGroupAndLoanType = " where l.group_id =? and glim.account_number="+glimAccount+" and l.loan_type_enum=4";
```
can we pass GLIM account number as we are passing group id, appending it in string doesn't look good.

nikpawar89 replied:
please look into this.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/group/api/GroupsApiResource.java:
```diff
@@ -508,4 +524,64 @@ public String postGroupTemplate(@FormDataParam("file") InputStream uploadedInput
+    public String retrieveglimAccounts(@PathParam("groupId") final Long groupId, @QueryParam("parentLoanAccountNo")final String parentLoanAccountNo,  @Context final UriInfo uriInfo)
+    {
+         this.context.authenticatedUser().validateHasReadPermission("GROUP");
+         List<GLIMContainer> glimContainer;
```
Try using  Collections.emptyList() to initialize container variables, it helps to avoid ugly null pointer exceptions

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java:
```diff
@@ -848,59 +862,121 @@ public String deleteLoanApplication(@PathParam("loanId") @ApiParam(value = "loan
     @ApiOperation(value = "Approve Loan Application | Recover Loan Guarantee | Undo Loan Application Approval | Assign a Loan Officer | Unassign a Loan Officer | Reject Loan Application | Applicant Withdraws from Loan Application | Disburse Loan Disburse Loan To Savings Account | Undo Loan Disbursal", httpMethod = "POST", notes = "Approve Loan Application:\n" + "Mandatory Fields: approvedOnDate\n ...
     @ApiImplicitParams({@ApiImplicitParam(value = "body", required = true, paramType = "body", dataType = "body", format = "body", dataTypeClass = LoansApiResourceSwagger.PostLoansLoanIdRequest.class)})
     @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoansApiResourceSwagger.PostLoansLoanIdResponse.class)})
-    public String stateTransitions(@PathParam("loanId") @ApiParam(value = "loanId") final Long loanId, @QueryParam("command") @ApiParam(value = "command") final String commandParam,
```
I think this swagger end point here was for Loan StateTransition, please switch the position of glimStateTransitions with stateTransitions. Later, you can add swagger end point for your API as well.

nikpawar89 replied:
could you please check this.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java:
```diff
@@ -848,59 +862,121 @@ public String deleteLoanApplication(@PathParam("loanId") @ApiParam(value = "loan
         }
 
+
+        if (result == null) { throw new UnrecognizedQueryParamException("command", commandParam); }
```
I think instead of throwing exception over here, we should let Client know about about incorrect param. 
By throwing exceptions, we sometimes reveal too much of inner mechanism to clients- just my thought.

nikpawar89 replied:
you can ignore this.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/GroupLoanIndividualMonitoringAccountData.java:
```diff
@@ -0,0 +1,119 @@
+    public String getLoanStatus() {
+        return loanStatus;
+    }
+
```
There are too many whitespaces and new line throughout the PR- just try to eliminate them, if possible.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:
```diff
@@ -145,6417 +145,6708 @@
-            final LoanProduct loanProduct, final Fund fund, final Staff officer, final CodeValue loanPurpose,
-            final LoanTransactionProcessingStrategy transactionProcessingStrategy,
-            final LoanProductRelatedDetail loanRepaymentScheduleDetail, final Set<LoanCharge> loanCharges,
-            final Set<LoanCollateral> collateral, final BigDecimal fixedEmiAmount, final List<LoanDisbursementDetails> disbursementDetails,
```
Please note, whenever we send a PR, it should contain either refactoring commit or new change commit. If the PR contains both then we should have first refactoring and then new feature commit.
I know its difficult, to judge initially whether the change you are going to make is going to be either refactoring or new change. But practicing such behavior makes it easy for reviewer. It also helps to understand the impact the PR is going to have on the existing functionality.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:
```diff
@@ -145,6417 +145,6708 @@
 
 @Entity
```
Note to self: in this mammoth file there are two effective changes : getter setter for loan type and GLIM object.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/data/CommandProcessingResultBuilder.java:
```diff
@@ -106,6 +108,16 @@ public CommandProcessingResultBuilder withProductId(final Long productId) {
         return this;
     }
 
+    public CommandProcessingResultBuilder withGsimId(final Long gsimId) {
```
This would have been painful during writing test cases and cause of many NPEs, I see gsimID and glimID added in commandProcessingResult result but in commandWrapper, we are mapping glimID and gsimID to entityID.
This has made it asymmetric and doesn't go well with Test infra. If possible take these new  instance variables out.

nikpawar89 replied:
For now you can keep this, we will revisit these two new params when we do more testing and see if refactoring is possible.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationCommandFromApiJsonHelper.java:
```diff
@@ -91,7 +91,9 @@
             LoanProductConstants.graceOnArrearsAgeingParameterName, LoanApiConstants.createStandingInstructionAtDisbursementParameterName,
-            LoanApiConstants.isTopup, LoanApiConstants.loanIdToClose, LoanApiConstants.datatables, LoanApiConstants.isEqualAmortizationParam, LoanProductConstants.ratesParamName));
+            LoanApiConstants.isTopup, LoanApiConstants.loanIdToClose, LoanApiConstants.datatables, LoanApiConstants.isEqualAmortizationParam,LoanProductConstants.ratesParamName,
+            LoanApiConstants.applicationId,                   // glim specific
```
If locking mechanism is added- applicationID can be removed altogether. Try the synchronization i suggested and omit this change if possible.

nikpawar89 replied:
let us revisit this when we have UI.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -365,11 +387,106 @@ public CommandProcessingResult submitApplication(final JsonCommand command) {
+                            {
+                                //**************Parent-Not an empty table********************
+                                accountNumber=this.accountNumberGenerator.generate(newLoanApplication, accountNumberFormat);
+                                newLoanApplication.updateAccountNo(accountNumber+"-1");
```
Including hyphens in account number would increase readability but some of the central banks won't like. 
I am aware of such incident, where one of the MF had account number with hyphen printed on cheque and clearing house refused to process it.

rrpawar96 replied:
Yes, hyphens have been removed.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -365,11 +387,106 @@ public CommandProcessingResult submitApplication(final JsonCommand command) {
+                            {
+                                //************** Parent-empty table********************
+
+                                accountNumber=this.accountNumberGenerator.generate(newLoanApplication, accountNumberFormat);
```
most of the stuff is repetitive while creating child account, try extracting function so it becomes readable and will reduce lines of code.

nikpawar89 replied:
we will revisit this later.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -365,11 +387,106 @@ public CommandProcessingResult submitApplication(final JsonCommand command) {
 
+
+            // loan account number generation
+            String accountNumber="";
```
Can we align GLIM creation with that of GSIM creation, I dont' see bulk creation in GLIM.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/GSIMReadPlatformServiceImpl.java:
```diff
@@ -0,0 +1,388 @@
+          List<SavingsSummaryCustom> childSavings;
+          for(GroupSavingsIndividualMonitoringAccountData gsimAccount:gsimInfo)
+          {
+               savingswhereClauseForGroup = " where sa.group_id = ? and sa.gsim_id="+gsimAccount.getGsimId()+" order by sa.status_enum ASC, sa.account_no ASC";
```
pass gsimID as object param.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsAccountReadPlatformServiceImpl.java:
```diff
@@ -789,7 +789,7 @@ public SavingsAccountTransactionsMapper() {
             sqlBuilder.append(" au.username as submittedByUsername, ");
             sqlBuilder.append(" nt.note as transactionNote, ") ;
-            sqlBuilder.append("tr.running_balance_derived as runningBalance, tr.is_reversed as reversed,");
+            sqlBuilder.append("tr.running_balance_derived as runningBalance, tr.is_reversed as reversed,tr.is_loan_disbursement as isLoanDisbursement,");
```
do we need this?- please remove if not.

nikpawar89 replied:
did you get a chance to look in to this?

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -86,470 +93,715 @@
+
+                                   accountNumber = this.accountNumberGenerator.generate(account, accountNumberFormat);
+                                   account.updateAccountNo(accountNumber + "-1");
+                                   gsimAccount = gsimWritePlatformService.addGSIMAccountInfo(accountNumber, group, BigDecimal.ZERO,
```
extract function of repetitive code.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -86,470 +93,715 @@
+                         .withClientId(account.clientId()) //
+                         .withGroupId(account.groupId()) //
+                         .withSavingsId(savingsId) //
+                        .withGsimId(gsimAccount==null ? 0 : gsimAccount.getId())
```
see if you can remove gsim id altogether from commandProcessingResult.

nikpawar89 commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -86,470 +93,715 @@
+                                   // Parent-Not an empty table
+
+                                   accountNumber = this.accountNumberGenerator.generate(account, accountNumberFormat);
+                                   account.updateAccountNo(accountNumber + "-1");
```
omit hyphens in account numbers.

francisguchie replied:
I agree with @nikpawar on removal of hyphens

francisguchie replied:
@nikpawar89 
Without hyphens the user can easily generate BBANs, issue check books to account holders 
standard account numbers should be 10 digits

rrpawar96 replied:
Yes, hyphens have been removed.

vorburger commented:
Please do not merge this PR as long as it contains the change to persistence.xml (more clearly seen in #777); I'll engage on https://jira.apache.org/jira/browse/FINERACT-902 re. that.

PS: Re. "Some checks were not successful" in "Build - Summary", please ignore, see [FINERACT-907](https://jira.apache.org/jira/browse/FINERACT-907).

rrpawar96 commented:
@vorburger, Separate commit has been created for review (where persistence.xml has been reverted, and fetchtype.Lazy has been added on two columns)

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:
```diff
@@ -140,11 +140,11 @@
     protected Group group;
 
-    @ManyToOne
+    @ManyToOne(fetch=FetchType.LAZY)
```
```suggestion
    @ManyToOne(fetch=FetchType.LAZY) // see FINERACT-902
```

vorburger replied:
@rrpawar96 did you want to add this?

rrpawar96 replied:
Yes @vorburger, I want to add this annotation and I will implement the fetchplan() in the future enhancement.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:
```diff
@@ -140,11 +140,11 @@
     protected Client client;
 
-    @ManyToOne(optional = true)
+    @ManyToOne(optional = true, fetch=FetchType.LAZY)
```
```suggestion
    @ManyToOne(optional = true, fetch=FetchType.LAZY)  // see FINERACT-902
```

vorburger replied:
@rrpawar96 did you want to add this?

rrpawar96 replied:
Yes

vorburger replied:
@rrpawar96 I think you still forgot this? I'll finally merge it anyway :smile: 

Perhaps you want to follow-up and add it in a subsequent PR.

vorburger reviewed (changes requested):
The FINERACT-902 problem (re. `persistence.xml`) is now OK for me here.  Re. the integration test build failures, could I encourage you to pro-actively drive that yourself, instead of waiting e.g. for me?  I've just raised #782 describing how I would like to see that work. -- I'll look at this PR again when you have managed to make it have a green build! (You can tag me on any PRs you raise proposing to ignore failing tests.)

nikpawar89 commented:
@rrpawar96  Besides changes suggested by @vorburger , I want you to address the previously discussed points : remove extra white spaces and lines.  Once this is done make sure you run the test cases and manually verify normal savings and loan application functionalities ( by installing Community App).  If regression is fine, I would give a go ahead from me to merge this PR. 

Having said that, as a Phase two( or in this phase itself- if required by other members) We have to add API documentation of this feature.

Once this is done, I would request you to take up Front end support for this feature. Any further bug fixes and enhancements can be addressed during this phase.

We are splitting this work in phases just to avoid the hassles of rebasing the code from time to time.

@vorburger @awasum  let me know what you guys think about this.

vorburger commented:
@rrpawar96 see https://github.com/apache/fineract#pull-requests Please do not wait for an active maintainer to come and help you re. build failures - it's your responsibility to get a green build. Have you checked the log, and done what the README suggests to do in such cases? :smiley_cat:

rrpawar96 commented:
Yes, @vorburger I have checked the log and noticed the failed IT but before I add the @ignore on flaky test cases, I was manually finding and fixing bugs which were occurred while creating savings and loan application because of this PR.  Now I am almost have done with it. 
I have noticed 2 failed flaky IT in build (https://travis-ci.org/github/apache/fineract/builds/680976084), out of that now one PR #790 already been created and another new PR #792 created for testInterestPostingForSavingsJobOutcome().

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountAssembler.java:
```diff
@@ -147,6 +152,20 @@ public SavingsAccount assembleFrom(final JsonCommand command, final AppUser subm
+             }
+        }
+
+       /* if(((Boolean)command.booleanPrimitiveValueOfParameterNamed("isGSIM"))!=null)
+        {
+             accountType = AccountType.GSIM;
+        }*/
```
?

rrpawar96 replied:
These comment lines has been removed.

vorburger commented:
@rrpawar96 I've just looked this over (again), and from my (technical) side there is nothing shocking here (anymore) that would prevent this from being merged. We now need others to do more reviews here - notably @nikpawar89 looks like he has previously reviewed, and should re-review this if you've incorporated his feedback. I'm now OK to merge this when we have LGTM from him.

nikpawar89 reviewed (changes requested):
@rrpawar96  all major concerns have been address, good work!
Please address these minor comments and we will merge this PR.

nikpawar89 commented:
Also, looks like you would have to rebase this PR, there is a conflict. 
Next steps: document API and start working on the Front end support.

rrpawar96 commented:
All necessary comments have been addressed and rebased, I think it's ready to get merged.

francisguchie commented:
Dear all,

I have shared a use case in the comments sections. I hope this helps for
others who are interested in this functionality

My Kind Regards

Francis Guchie Kirago
*Skype:* francisguchie
*Whatsapp: *232 79 19 44 07
*LINKEDIN:* https://www.linkedin.com/in/francis-guchie-kirago-a4379617/
twitter: @FrancisGuchie









On Mon, May 4, 2020 at 1:56 PM Rahul Pawar <notifications@github.com> wrote:

> All necessary comments have been addressed and rebased, I think it's ready
> to get merged.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/738#issuecomment-623479143>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFNCABXNKUIT44UQFI5M5I3RP3CPRANCNFSM4LPITBVA>
> .
>

nikpawar89 commented:
Looks good to me now @rrpawar96 .  Kindly create API documentations and start looking at the Front end.
 @vorburger  I am fine with these changes, if you are satisfied we should merge this PR.

nikpawar89 commented:
> Dear all, I have shared a use case in the comments sections. I hope this helps for others who are interested in this functionality My Kind Regards Francis Guchie Kirago *Skype:* francisguchie *Whatsapp: *232 79 19 44 07 *LINKEDIN:* https://www.linkedin.com/in/francis-guchie-kirago-a4379617/ twitter: @francisguchie
> […](#)
> On Mon, May 4, 2020 at 1:56 PM Rahul Pawar ***@***.***> wrote: All necessary comments have been addressed and rebased, I think it's ready to get merged. — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#738 (comment)](https://github.com/apache/fineract/pull/738#issuecomment-623479143)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFNCABXNKUIT44UQFI5M5I3RP3CPRANCNFSM4LPITBVA> .

@francisguchie  thanks for your time, would you like to volunteer for QA?
Also, there is going to be second phase of development for this feature, once we UI support.

vorburger commented:
@rrpawar96 sorry to be a PITA but could I suggest you change the message of the (now single, thanks) commit from "GLIM_GSIM enhancement" to "GLIM_GSIM enhancement (FINERACT-603)"? Just do:

    git commit --amend -m "GLIM_GSIM enhancement (FINERACT-603)"
    git push --force rrpawar96

 > nikpawar89 : @vorburger I am fine with these changes, if you are satisfied we should merge this PR.

Ok, great, then I'll finally merge this one ASAP.

vorburger commented:
FINERACT-603 comment says functional verification will be performed by @francisguchie post merge.

francisguchie commented:
I would like to thank @vorburger @nikpawar89  i have seen the merge. 
time to do functional review. We need the UI as well @nikpawar89  thank you for promising to atend to this too. 

In the meantime i will use what i have available as i await the UI updates

francisguchie commented:
@nikpawar89  
Yes, I would like to volunteer for the QA

nikpawar89 commented:
@francisguchie Thank you for contributing the use cases. We are working on the getting UI to be rebased on our latest community app code -so that it will become easier for you to do functional testing. I will keep you posted.
